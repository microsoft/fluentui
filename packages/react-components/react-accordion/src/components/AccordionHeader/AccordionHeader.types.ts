import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import type { ARIAButtonSlotProps } from '@fluentui/react-aria';

export type AccordionHeaderSize = 'small' | 'medium' | 'large' | 'extra-large';
export type AccordionHeaderExpandIconPosition = 'start' | 'end';

export type AccordionHeaderContextValue = Required<Pick<AccordionHeaderProps, 'expandIconPosition' | 'size'>> & {
  disabled: boolean;
  open: boolean;
};

export type AccordionHeaderContextValues = {
  accordionHeader: AccordionHeaderContextValue;
};

export type AccordionHeaderSlots = {
  /**
   * The element wrapping the button. By default this is a div, but can be a heading.
   */
  root: Slot<'div', 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'>;
  /**
   * The component to be used as button in heading
   */
  button: NonNullable<Slot<ARIAButtonSlotProps<'a'>>>;
  /**
   * Expand icon slot rendered before (or after) children content in heading.
   */
  expandIcon: Slot<'span'>;
  /**
   * Expand icon slot rendered before (or after) children content in heading.
   */
  icon?: Slot<'div'>;
};

export type AccordionHeaderProps = ComponentProps<Partial<AccordionHeaderSlots>> & {
  /**
   * The position of the expand  icon slot in heading.
   */
  expandIconPosition?: AccordionHeaderExpandIconPosition;

  /**
   * Indicates if the AccordionHeader should be rendered inline.
   */
  inline?: boolean;

  /**
   * Size of spacing in the heading.
   */
  size?: AccordionHeaderSize;
};

export type AccordionHeaderState = ComponentState<AccordionHeaderSlots> &
  Required<Pick<AccordionHeaderProps, 'inline'>> &
  AccordionHeaderContextValue;
