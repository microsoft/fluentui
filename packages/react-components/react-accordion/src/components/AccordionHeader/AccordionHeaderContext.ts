import * as React from 'react';
import type { AccordionHeaderContextValue } from './AccordionHeader.types';

export const AccordionHeaderContext = React.createContext<AccordionHeaderContextValue | undefined>(undefined);

const accordionHeaderContextDefaultValue = {
  open: false,
  disabled: false,
  size: 'medium',
  expandIconPosition: 'start',
};

export const useAccordionHeaderContext = () =>
  React.useContext(AccordionHeaderContext) ?? accordionHeaderContextDefaultValue;
