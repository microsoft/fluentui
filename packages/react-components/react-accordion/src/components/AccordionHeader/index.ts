export * from './AccordionHeader';
export * from './AccordionHeader.types';
export * from './renderAccordionHeader';
export * from './useAccordionHeader';
export * from './useAccordionHeaderContextValues';
export * from './useAccordionHeaderStyles.styles';
