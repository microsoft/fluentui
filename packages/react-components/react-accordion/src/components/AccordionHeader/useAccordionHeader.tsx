import * as React from 'react';
import { getNativeElementProps, resolveShorthand, useEventCallback } from '@fluentui/react-utilities';
import { useAccordionItemContext_unstable } from '../AccordionItem/index';
import { useARIAButtonShorthand } from '@fluentui/react-aria';
import type { AccordionHeaderProps, AccordionHeaderState } from './AccordionHeader.types';
import { useAccordionContext_unstable } from '../Accordion/AccordionContext';
import { ChevronRightRegular } from '@fluentui/react-icons';
import { useFluent_unstable as useFluent } from '@fluentui/react-shared-contexts';

/**
 * Returns the props and state required to render the component
 * @param props - AccordionHeader properties
 * @param ref - reference to root HTMLElement of AccordionHeader
 */
export const useAccordionHeader_unstable = (
  props: AccordionHeaderProps,
  ref: React.Ref<HTMLElement>,
): AccordionHeaderState => {
  const { as, icon, button, expandIcon, inline = false, size = 'medium', expandIconPosition = 'start' } = props;
  const { onHeaderClick: onAccordionHeaderClick, disabled, open } = useAccordionItemContext_unstable();

  /**
   * force disabled state on button if accordion isn't collapsible
   * and this is the only item opened
   */
  const disabledFocusable = useAccordionContext_unstable(ctx => !ctx.collapsible && ctx.openItems.length === 1 && open);

  const { dir } = useFluent();

  // Calculate how to rotate the expand icon [>] (ChevronRightRegular)
  let expandIconRotation: 0 | 90 | -90 | 180;
  if (expandIconPosition === 'end') {
    // If expand icon is at the end, the chevron points up [^] when open, and down [v] when closed
    expandIconRotation = open ? -90 : 90;
  } else {
    // Otherwise, the chevron points down [v] when open, and right [>] (or left [<] in RTL) when closed
    expandIconRotation = open ? 90 : dir !== 'rtl' ? 0 : 180;
  }

  const buttonShorthand = useARIAButtonShorthand(button, {
    required: true,
    defaultProps: {
      disabled,
      disabledFocusable,
      'aria-expanded': open,
      type: 'button',
    },
  });

  return {
    disabled,
    open,
    size,
    inline,
    expandIconPosition,
    components: {
      root: 'div',
      button: 'button',
      expandIcon: 'span',
      icon: 'div',
    },
    root: getNativeElementProps(as || 'div', {
      ref,
      ...props,
    }),
    icon: resolveShorthand(icon),
    expandIcon: resolveShorthand(expandIcon, {
      required: true,
      defaultProps: {
        children: <ChevronRightRegular style={{ transform: `rotate(${expandIconRotation}deg)` }} />,
        'aria-hidden': true,
      },
    }),
    button: {
      ...buttonShorthand,
      onClick: useEventCallback(
        (ev: React.MouseEvent<HTMLButtonElement & HTMLDivElement & HTMLSpanElement & HTMLAnchorElement>) => {
          buttonShorthand.onClick?.(ev);
          if (!ev.defaultPrevented) {
            onAccordionHeaderClick(ev);
          }
        },
      ),
    },
  };
};
