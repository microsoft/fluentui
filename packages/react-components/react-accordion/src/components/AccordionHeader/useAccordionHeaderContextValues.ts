import * as React from 'react';
import type { AccordionHeaderState, AccordionHeaderContextValues } from './AccordionHeader.types';
import type { AccordionHeaderContextValue } from '../../contexts/accordionHeader';

export function useAccordionHeaderContextValues_unstable(state: AccordionHeaderState): AccordionHeaderContextValues {
  const { disabled, expandIconPosition, open, size } = state;

  const accordionHeader = React.useMemo<AccordionHeaderContextValue>(
    () => ({
      disabled,
      expandIconPosition,
      open,
      size,
    }),
    [disabled, expandIconPosition, open, size],
  );

  return { accordionHeader };
}
