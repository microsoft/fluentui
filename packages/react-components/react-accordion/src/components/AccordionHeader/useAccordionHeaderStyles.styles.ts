import { shorthands, makeStyles, mergeClasses } from '@griffel/react';
import { createFocusOutlineStyle } from '@fluentui/react-tabster';
import { tokens, typographyStyles } from '@fluentui/react-theme';
import type { AccordionHeaderSlots, AccordionHeaderState } from './AccordionHeader.types';
import type { SlotClassNames } from '@fluentui/react-utilities';

export const accordionHeaderClassNames: SlotClassNames<AccordionHeaderSlots> = {
  root: 'fui-AccordionHeader',
  button: 'fui-AccordionHeader__button',
  expandIcon: 'fui-AccordionHeader__expandIcon',
  icon: 'fui-AccordionHeader__icon',
};

const useStyles = makeStyles({
  // TODO: this should be extracted to another package
  resetButton: {
    boxSizing: 'content-box',
    backgroundColor: 'inherit',
    color: 'inherit',
    fontFamily: 'inherit',
    fontSize: 'inherit',
    lineHeight: 'normal',
    ...shorthands.overflow('visible'),
    ...shorthands.padding(0),
    WebkitAppearance: 'button',
    textAlign: 'unset',
  },
  focusIndicator: createFocusOutlineStyle(),
  root: {
    color: tokens.colorNeutralForeground1,
    backgroundColor: tokens.colorTransparentBackground,
    ...shorthands.margin(0),
    ...shorthands.borderRadius(tokens.borderRadiusMedium),
  },
  rootDisabled: {
    backgroundImage: 'none',
    color: tokens.colorNeutralForegroundDisabled,
  },
  rootInline: {
    display: 'inline-block',
  },
  button: {
    position: 'relative',
    width: '100%',
    ...shorthands.border('1px', 'solid', 'transparent'),
    ...shorthands.padding(0, tokens.spacingHorizontalM, 0, tokens.spacingHorizontalMNudge),
    minHeight: '44px',
    display: 'flex',
    alignItems: 'center',
    cursor: 'pointer',
    ...typographyStyles.body1,
    boxSizing: 'border-box',
  },
  buttonSmall: {
    minHeight: '32px',
    fontSize: tokens.fontSizeBase200,
  },
  buttonLarge: {
    lineHeight: tokens.lineHeightBase400,
    fontSize: tokens.fontSizeBase400,
  },
  buttonExtraLarge: {
    lineHeight: tokens.lineHeightBase500,
    fontSize: tokens.fontSizeBase500,
  },
  buttonInline: {
    display: 'inline-flex',
  },
  buttonExpandIconEndNoIcon: {
    paddingLeft: tokens.spacingHorizontalM,
  },
  buttonExpandIconEnd: {
    paddingRight: tokens.spacingHorizontalMNudge,
  },
  expandIcon: {
    height: '100%',
    display: 'flex',
    alignItems: 'center',
    lineHeight: tokens.lineHeightBase500,
    fontSize: tokens.fontSizeBase500,
  },
  expandIconStart: {
    paddingRight: tokens.spacingHorizontalS,
  },
  expandIconEnd: {
    flexGrow: 1,
    flexShrink: 1,
    flexBasis: '0%',
    display: 'flex',
    justifyContent: 'flex-end',
    paddingLeft: tokens.spacingHorizontalS,
  },
  icon: {
    height: '100%',
    display: 'flex',
    alignItems: 'center',
    paddingRight: tokens.spacingHorizontalS,
    lineHeight: tokens.lineHeightBase500,
    fontSize: tokens.fontSizeBase500,
  },
});

/** Applies style classnames to slots */
export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
  const styles = useStyles();
  state.root.className = mergeClasses(
    accordionHeaderClassNames.root,
    styles.root,
    state.inline && styles.rootInline,
    state.disabled && styles.rootDisabled,
    state.root.className,
  );

  state.button.className = mergeClasses(
    accordionHeaderClassNames.button,
    styles.resetButton,
    styles.button,
    styles.focusIndicator,
    state.expandIconPosition === 'end' && !state.icon && styles.buttonExpandIconEndNoIcon,
    state.expandIconPosition === 'end' && styles.buttonExpandIconEnd,
    state.inline && styles.buttonInline,
    state.size === 'small' && styles.buttonSmall,
    state.size === 'large' && styles.buttonLarge,
    state.size === 'extra-large' && styles.buttonExtraLarge,
    state.button.className,
  );

  if (state.expandIcon) {
    state.expandIcon.className = mergeClasses(
      accordionHeaderClassNames.expandIcon,
      styles.expandIcon,
      state.expandIconPosition === 'start' && styles.expandIconStart,
      state.expandIconPosition === 'end' && styles.expandIconEnd,
      state.expandIcon.className,
    );
  }
  if (state.icon) {
    state.icon.className = mergeClasses(accordionHeaderClassNames.icon, styles.icon, state.icon.className);
  }
  return state;
};
