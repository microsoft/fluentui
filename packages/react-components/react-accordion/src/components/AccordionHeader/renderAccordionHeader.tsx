/** @jsxRuntime classic */
/** @jsx createElement */

import { createElement } from '@fluentui/react-jsx-runtime';

import { getSlotsNext } from '@fluentui/react-utilities';
import { AccordionHeaderContext } from './AccordionHeaderContext';
import type { AccordionHeaderState, AccordionHeaderSlots, AccordionHeaderContextValues } from './AccordionHeader.types';

/**
 * Function that renders the final JSX of the component
 */
export const renderAccordionHeader_unstable = (
  state: AccordionHeaderState,
  contextValues: AccordionHeaderContextValues,
) => {
  const { slots, slotProps } = getSlotsNext<AccordionHeaderSlots>(state);

  return (
    <AccordionHeaderContext.Provider value={contextValues.accordionHeader}>
      <slots.root {...slotProps.root}>
        <slots.button {...slotProps.button}>
          {state.expandIconPosition === 'start' && <slots.expandIcon {...slotProps.expandIcon} />}
          {slots.icon && <slots.icon {...slotProps.icon} />}
          {slotProps.root.children}
          {state.expandIconPosition === 'end' && <slots.expandIcon {...slotProps.expandIcon} />}
        </slots.button>
      </slots.root>
    </AccordionHeaderContext.Provider>
  );
};
