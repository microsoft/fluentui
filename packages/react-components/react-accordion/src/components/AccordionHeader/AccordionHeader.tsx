import * as React from 'react';
import { useAccordionHeader_unstable } from './useAccordionHeader';
import { renderAccordionHeader_unstable } from './renderAccordionHeader';
import { useAccordionHeaderStyles_unstable } from './useAccordionHeaderStyles';
import { useAccordionHeaderContextValues_unstable } from './useAccordionHeaderContextValues';
import type { AccordionHeaderProps } from './AccordionHeader.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * Define a styled AccordionHeader, using the `useAccordionHeader_unstable` and `useAccordionHeaderStyles_unstable`
 * hooks.
 */
export const AccordionHeader: ForwardRefComponent<AccordionHeaderProps> = React.forwardRef((props, ref) => {
  const state = useAccordionHeader_unstable(props, ref);
  const contextValues = useAccordionHeaderContextValues_unstable(state);

  useAccordionHeaderStyles_unstable(state);

  return renderAccordionHeader_unstable(state, contextValues);
});

AccordionHeader.displayName = 'AccordionHeader';
