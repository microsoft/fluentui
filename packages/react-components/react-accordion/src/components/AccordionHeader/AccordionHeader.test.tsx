import { resetIdsForTests } from '@fluentui/react-utilities';
import * as React from 'react';
import { AccordionHeader } from './AccordionHeader';
import { AccordionHeaderProps } from './AccordionHeader.types';
import * as renderer from 'react-test-renderer';
import { isConformant } from '../../testing/isConformant';
import { Accordion } from '../Accordion/Accordion';
import { AccordionItem } from '../AccordionItem';
import { AccordionPanel } from '../AccordionPanel';

describe('AccordionHeader', () => {
  isConformant<AccordionHeaderProps>({
    Component: AccordionHeader,
    displayName: 'AccordionHeader',
    testOptions: {
      'has-static-classnames': [
        {
          props: {
            icon: 'Test Icon',
          },
        },
      ],
    },
  });

  afterEach(() => {
    resetIdsForTests();
  });

  /**
   * Note: see more visual regression tests for AccordionHeader in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<AccordionHeader>Default AccordionHeader</AccordionHeader>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should invoke click and toggle', () => {
    const mockClick = jest.fn();
    const component = renderer.create(
      <Accordion collapsible openItems={0} onToggle={mockClick}>
        <AccordionItem value={0}>
          <AccordionHeader button={{ onClick: mockClick }}>Header</AccordionHeader>
          <AccordionPanel>Panel</AccordionPanel>
        </AccordionItem>
      </Accordion>,
    );
    renderer.act(() => {
      component.root.findAllByType('button')[0].props.onClick({ defaultPrevented: false });
    });
    expect(mockClick).toBeCalledTimes(2);
  });

  it('should invoke click and prevent toggle', () => {
    const mockClick = jest.fn();
    const component = renderer.create(
      <Accordion collapsible openItems={0} onToggle={mockClick}>
        <AccordionItem value={0}>
          <AccordionHeader button={{ onClick: mockClick }}>Header</AccordionHeader>
          <AccordionPanel>Panel</AccordionPanel>
        </AccordionItem>
      </Accordion>,
    );
    renderer.act(() => {
      component.root.findAllByType('button')[0].props.onClick({ defaultPrevented: true });
    });
    expect(mockClick).toBeCalledTimes(1);
  });
});
