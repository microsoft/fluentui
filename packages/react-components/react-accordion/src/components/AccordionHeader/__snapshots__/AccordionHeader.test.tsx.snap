// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionHeader renders a default state 1`] = `
<div
  className="fui-AccordionHeader"
>
  <button
    aria-expanded={false}
    className="fui-AccordionHeader__button"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span
      aria-hidden={true}
      className="fui-AccordionHeader__expandIcon"
    >
      <svg
        aria-hidden={true}
        className=""
        fill="currentColor"
        height="1em"
        style={
          Object {
            "transform": "rotate(0deg)",
          }
        }
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.65 4.15c.2-.2.5-.2.7 0l5.49 5.46c.21.22.21.57 0 .78l-5.49 5.46a.5.5 0 01-.7-.7L12.8 10 7.65 4.85a.5.5 0 010-.7z"
          fill="currentColor"
        />
      </svg>
    </span>
    Default AccordionHeader
  </button>
</div>
`;
